execution/tvf/percentile_watermark_generator.go

```diff
@@ -0,0 +1,297 @@
+	timeField  octosql.VariableName
+	events     int
+	percentile int
+	frequency  int
+}
+
+var (
+	percentileWatermarkPrefix = []byte("$percentile_watermark$")
+
+	percentileWatermarkEventsPrefix      = []byte("$percentile_watermark_events$")
+	percentileWatermarkEventsCountPrefix = []byte("$percentile_watermark_events_count$")
+	percentileWatermarkEventsSeenPrefix  = []byte("$percentile_watermark_events_seen$")
+)
+
+func (s *PercentileWatermarkGeneratorStream) GetWatermark(ctx context.Context, tx storage.StateTransaction) (time.Time, error) {
+	watermarkStorage := storage.NewValueState(tx.WithPrefix(percentileWatermarkPrefix))
+
+	var currentWatermark octosql.Value
+	err := watermarkStorage.Get(&currentWatermark)
+	if err == storage.ErrNotFound {
+		currentWatermark = octosql.MakeTime(time.Time{})
+	} else if err != nil {
+		return time.Time{}, errors.Wrap(err, "couldn't get current watermark from storage")
+	}
+
+	return currentWatermark.AsTime(), nil
+}
+
+func (s *PercentileWatermarkGeneratorStream) Next(ctx context.Context) (*execution.Record, error) {
+	srcRecord, err := s.source.Next(ctx)
+	if err != nil {
+		if err == execution.ErrEndOfStream {
+			return nil, execution.ErrEndOfStream
+		}
+		return nil, errors.Wrap(err, "couldn't get source record")
+	}
+
+	timeValue := srcRecord.Value(s.timeField)
+	if timeValue.GetType() != octosql.TypeTime {
+		return nil, fmt.Errorf("couldn't get time field '%v' as time, got: %v", s.timeField.String(), srcRecord.Value(s.timeField))
+	}
+
+	tx := storage.GetStateTransactionFromContext(ctx)
+	watermarkStorage := storage.NewValueState(tx.WithPrefix(percentileWatermarkPrefix))
+	eventsStorage := storage.NewDeque(tx.WithPrefix(percentileWatermarkEventsPrefix))
+	eventsCountStorage := storage.NewMap(tx.WithPrefix(percentileWatermarkEventsCountPrefix))
+	eventsSeenStorage := storage.NewValueState(tx.WithPrefix(percentileWatermarkEventsSeenPrefix))
+
+	// Extracting eventsSeen value (remember: this equals to number of events seen before watermark update)
+	var eventsSeen octosql.Value
+	err = eventsSeenStorage.Get(&eventsSeen)
+	if err == storage.ErrNotFound {
+		eventsSeen = octosql.MakeInt(0)
+	} else if err != nil {
+		return nil, errors.Wrap(err, "couldn't get events seen count")
+	}
+	eventsSeen = octosql.MakeInt(eventsSeen.AsInt() + 1) // Adding +1 as we've just extracted next record from source
+
+	// Adding newest event
+	err = addNewestEvent(timeValue, tx)
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't add newest event to storage")
+	}
+
+	eventsLength, err := eventsStorage.Length()
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't get events deque length")
+	}
+
+	// In this situation we need to remove oldest event => this doesn't occur only when the oldest event seen is the first ever and length = 5 (update watermark but DON't remove oldest one)
+	if eventsLength > s.events { // TODO - which is more readable: this or 'eventsLength == s.events + 1' ?
```

Comment:
I think this one is more readable, since it suggests that if we have too many events stored, we get rid of the oldest one stored :)